compiler/semantic/checker.go

```diff
@@ -0,0 +1,1081 @@
+	var elemType super.Type
+	switch typ := super.TypeUnder(containerType).(type) {
+	case *super.TypeArray:
+		elemType = typ.Type
+	case *super.TypeSet:
+		elemType = typ.Type
+	case *super.TypeError:
+		if isUnknown(typ) {
+			elemType = c.unknown
+		}
+	}
+	return elemType
+}
+
+func (c *checker) arrayElems(typ super.Type, elems []sem.ArrayElem) super.Type {
+	fuser := c.newFuser()
+	for _, elem := range elems {
+		switch elem := elem.(type) {
+		case *sem.SpreadElem:
+			fuser.fuse(c.expr(typ, elem.Expr))
+		case *sem.ExprElem:
+			fuser.fuse(c.expr(typ, elem.Expr))
+		default:
+			panic(elem)
+		}
+	}
+	return fuser.Type(c)
+}
+
+func (c *checker) recordElems(typ super.Type, elems []sem.RecordElem) super.Type {
+	fuser := c.newFuser()
+	for _, elem := range elems {
+		switch elem := elem.(type) {
+		case *sem.SpreadElem:
+			fuser.fuse(c.expr(typ, elem.Expr))
+		case *sem.FieldElem:
+			column := super.Field{Name: elem.Name, Type: c.expr(typ, elem.Value)}
+			fuser.fuse(c.t.sctx.MustLookupTypeRecord([]super.Field{column}))
+		default:
+			panic(elem)
+		}
+	}
+	return fuser.Type(c)
+}
+
+func (c *checker) callBuiltin(call *sem.CallExpr, args []super.Type) super.Type {
+	// XXX This will be handled in a subsequent PR where we add type signatures
+	// to the package containing the built-in function implementatons.
+	return c.unknown
+}
+
+func (c *checker) callFunc(call *sem.CallExpr, args []super.Type) super.Type {
+	f := c.funcs[call.Tag]
+	if len(args) != len(f.Params) {
+		// The translator has already checked that len(args) is len(params)
+		// but when there's an error, mismatches can still show up here so
+		// we ignore these here.
+		return c.unknown
+	}
+	fields := make([]super.Field, 0, len(args))
+	for k, param := range f.Params {
+		fields = append(fields, super.Field{Name: param, Type: args[k]})
+	}
+	argsType := c.t.sctx.MustLookupTypeRecord(fields)
+	if typ, ok := c.checked[argsType]; ok {
+		return typ
+	}
+	// For recursive calls, we use unknown for the base type to halt the recursion
+	// then fill the actual type computed from the unknown base type.  This has the
+	// downside that we now have an error unknown in the sum type of the return value
+	// of all recursive functions.  When we add (optional) type signatures to functions,
+	// this problem will (partially) go away.
+	c.checked[argsType] = c.unknown
+	typ := c.expr(argsType, f.Body)
+	c.checked[argsType] = typ
+	return typ
+}
+
+type pathType struct {
+	elems []string
+	typ   super.Type
+}
+
+func (c *checker) pathsToType(paths []pathType) super.Type {
+	fuser := c.newFuser()
+	for _, path := range paths {
+		fuser.fuse(c.pathToRec(path.typ, path.elems))
+	}
+	return fuser.Type(c)
+}
+
+func (c *checker) pathToRec(typ super.Type, elems []string) super.Type {
+	for ; len(elems) > 0; elems = elems[:len(elems)-1] {
+		last := elems[len(elems)-1]
+		typ = c.t.sctx.MustLookupTypeRecord([]super.Field{{Name: last, Type: typ}})
```

Comment:
```suggestion
	```suggestion
	for _, elem := range slices.Backward(elems) {
typ}})
		typ = c.t.sctx.MustLookupTypeRecord([]super.Field{{Name: elem, Type: typ}})
```